### PR TITLE
Adding oecryptoopenssl_3 target and updating samples to build with OpenSSL 3.0

### PR DESCRIPTION
**Pull request by dcarpente**
This pull request contains the following changes:

1. Create new oecryptoopenssl_3 target
2. Update enclave/crypto/openssl/init.c to initialize based on the version of OpenSSL
3. Update/Fix openssl_3 target to install OpenSSL 3.0 public headers to 3rdparty/openssl_3/openssl folder
4. Create copies of attested_cert_apis, attestation_plugin_cert, crypto/enclave, sealKey, tls_e2e tests that build with openssl_3
    a. Increase NumHeapPages for attested_cert_apis, attestation_plugin_cert, crypto/enclave to avoid memory issues when compiling with snmalloc
5. Update all pkgconfig files, READMEs, and sample Makefiles to work with openssl_3
    a. Update add-samples.cmake to run attested-tls and file-encryptor samples with openssl_3
6. Force include define_getpid_noop.h in OpenSSL source to overwrite usages of getpid(), which isn't supported in openenclave. (See https://github.com/openssl/openssl/issues/21139 for more context.)
7. Fix issue where append-unsupported would try to append to OpenSSL 3.0 header prov_ssl.h

### CODE REVIEW

**Comment by dcarpente** [comment]
bors try

File: tests/openssl/enc/enc.c [hunk 1]
```diff
@@ -26,10 +26,6 @@ int enc_test(int argc, char** argv, char** env)
     /* Directly use environ from host. */
     __environ = env;
```
Comment (radhikaj) [inline tests/openssl/enc/enc.c:28]: why was this removed?

Comment (dcarpente) [reply tests/openssl/enc/enc.c:28]: This was moved to enclave/crypto/openssl/init.c! When I made the original openssl_3 change I wanted to avoid making changes to enclave/crypto yet.

File: pkgconfig/README.md [hunk 2]
```diff
@@ -49,12 +48,28 @@ cxxflags=`pkg-config oeenclave-clang++ --cflags`
 libs=`pkg-config oeenclave-clang++ --libs`
 $ clang++-11 ${cxxflags} -o enc enc.cpp ${libs}
 ```
-**Note**:  While `clang++-11` is our supported and recommended version, enclave applications may be built with `clang++-10` as well.
```
Comment (radhikaj) [inline pkgconfig/README.md:52]: thanks for removing!

File: pkgconfig/README.md [hunk 3]
```diff
@@ -49,12 +48,28 @@ cxxflags=`pkg-config oeenclave-clang++ --cflags`
...
+To build an enclave application with a specific crypto library, additional commands are needed to incorporate additional source files and headers. In this example, `OE_CRYPTO_LIB` is set to `openssl`. The values that `OE_CRYPTO_LIB` supports are:
+- `mbedtls`
+- `openssl`
+- `openssl_symcrypt_fips`
+- `openssl_3`
+
+```
+OE_CRYPTO_LIB=openssl
+cxxflags=`pkg-config oeenclave-clang++ --cflags`
+cryptoflags=`pkg-config oeenclave-clang++ --variable=${OE_CRYPTO_LIB}flags
```
Comment (radhikaj) [inline pkgconfig/README.md:61]: nice!

File: samples/pluggable_allocator/enclave/Makefile [hunk 4]
```diff
@@ -33,8 +34,8 @@ build:
 	oeedger8r ../allocator_demo.edl --trusted \
 		--search-path $(INCDIR) \
 		--search-path $(INCDIR)/openenclave/edl/sgx
-	$(CXX) -g -c $(CXXFLAGS) -DOE_API_VERSION=2 enc.cpp -o enc.o
-	$(CC) -g -c $(CFLAGS) -DOE_API_VERSION=2 allocator_demo_t.c -o allocator_demo_t.o
+	$(CXX) -g -c $(CRYPTO_FLAGS) $(CXXFLAGS) -DOE_API_VERSION=2 enc.cpp -o enc.o
```
Comment (radhikaj) [inline samples/pluggable_allocator/enclave/Makefile:37]: One issue - this is a breaking change to existing apps. We should flag it as such! They would all definitely need to change it. I wonder if there is any way around this

Comment (dcarpente) [reply samples/pluggable_allocator/enclave/Makefile:37]: "CRYPTO_FLAGS" is an empty string for all crypto libs except openssl_3. Existing apps can continue building with openssl/mbedtls/etc and ignoring CRYPTO_FLAGS and should not be affected.

But this still may be confusing and/or not clearly presented, and I'm open to suggestions on a more intuitive or better explained way.

Comment (radhikaj) [reply samples/pluggable_allocator/enclave/Makefile:37]: Is there a way to push CRYPTO_FLAGS into  $(CXXFLAGS)(or CFLAGS)?
Like CXXFLAGS = CXXFLAGS + CRYPTO_FLAGS somewhere?

Comment (dcarpente) [reply samples/pluggable_allocator/enclave/Makefile:37]: We can absolutely do CXXFLAGS = CRYPTO_FLAGS + CXXFLAGS in the sample Makefiles. I'll make that change.

The problem is that, regardless, customers will need to pull "--variable=${OE_CRYPTO_LIB}flags" info from pkg-config to build with openssl_3. We can't add the additional includes to pkg-config --cflags because they are different for openssl_3 vs openssl

Comment (CyanDevs) [reply samples/pluggable_allocator/enclave/Makefile:37]: Could we put together a guidance for customers upgrading to OpenSSL 3? And, Is there anything customers need to do if not upgrading?

Comment (dcarpente) [reply samples/pluggable_allocator/enclave/Makefile:37]: To you have advise where to put that guidance? Maybe a new section in docs/OpenSSLSupport.md?

The way this current change is now, customers who don't upgrade should not have to do anything.

File: 3rdparty/openssl/CMakeLists.txt [hunk 5]
```diff
@@ -1360,6 +1360,22 @@ list(
...
+# which uses getpid that is not supported on Windows.
+# We force include this header so that openssl_get_fork_id returns 0 instead.
+if (WIN32)
+  string(REPLACE "/" "\\" COMPILE_CURRENT_SOURCE_DIR
+                 "${CMAKE_CURRENT_SOURCE_DIR}")
+  set(DEFINE_GETPID_COMPILE_OPTION
+      "-include${COMPILE_CURRENT_SOURCE_DIR}\\define_getpid_noop.h")
+else ()
+  set(DEFINE_GETPID_COMPILE_OPTION
+      "-include${CMAKE_CURRENT_SOURCE_DIR}/define_getpid_noop.h")
```
[the remark this answers is not in the document]
Comment (dcarpente) [reply 3rdparty/openssl/CMakeLists.txt:1373]: We don't need to include define_getpid_noop.h for enclaves running on Linux. But I wasn't sure if there were cases where an enclave compiled on Linux can be run on Windows?

[the remark this answers is not in the document]
Comment (CyanDevs) [reply 3rdparty/openssl/CMakeLists.txt:1373]: That is a valid case, so good point. Thanks.

**Comment by dcarpente** [comment]
bors try

File: samples/pluggable_allocator/enclave/Makefile (hunk 4 above)
Comment (dcarpente) [reply samples/pluggable_allocator/enclave/Makefile]: I reformatted the Makefiles to concatenate CRYPTO_FLAGS with CFLAGS. I also added a new section in docs/OpenSSLSupport.md with some pointers on building enclaves with OpenSSL 3.0.

**Review by CyanDevs — APPROVED** [review]
LGTM

**Comment by dcarpente** [comment]
I've squashed the commits. Thanks for your review @CyanDevs and @radhikaj.

File: pkgconfig/oeenclave-clang.pc [hunk 6]
```diff
@@ -5,13 +5,17 @@ prefix=${pcfiledir}/../..
...
 openssllibs=@ENCLAVE_OPENSSL_CRYPTO_LIBS@
+openssl_3libs=@ENCLAVE_OPENSSL_3_CRYPTO_LIBS@
 openssl_symcrypt_fipslibs=@ENCLAVE_OPENSSL_SYMCRYPT_FIPS_CRYPTO_LIBS@
 
+openssl_3flags=@ENCLAVE_OPENSSL_3_INCLUDES@
+
 Name: Open Enclave
 Description: Open Enclave clang flags and libraries for building enclaves.
 Version: @OE_VERSION@
 Requires:
```
Comment (radhikaj) [inline pkgconfig/oeenclave-clang.pc:19]: I think you need to walk me through this. What is the value of @ENCLAVE_OPENSSL_3_CRYPTO_LIBS@?

**Review by mingweishih — APPROVED** [review]
LGTM

File: pkgconfig/oeenclave-clang.pc (hunk 6 above)
Comment (dcarpente) [reply pkgconfig/oeenclave-clang.pc:19]: That value is defined in pkgconfig/CMakeLists.txt.
@ENCLAVE_OPENSSL_3_CRYPTO_LIBS@ is "-loecryptoopenssl_3 -lopensslssl_3 -lopensslcrypto_3". These are build arguments to add libraries containing the OpenSSL 3.0 source code.
@ENCLAVE_OPENSSL_3_INCLUDES@ is "-I\${includedir}/openenclave/3rdparty/openssl_3". This is a build argument to add an include directory that contains OpenSSL 3.0 header files.

**Review by radhikaj — APPROVED** [review]
Thanks Danny!

**Comment by dcarpente** [comment]
bors try

File: 3rdparty/openssl/define_getpid_noop.h [hunk 7]
```diff
@@ -0,0 +1,11 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+// We force include this header on threads_pthread.c to overwrite getpid(),
+// which isn't supported on Windows.
+#define getpid oe_openssl_get_pid
+
+int oe_openssl_get_pid()
```
Comment (anakrish) [inline 3rdparty/openssl/define_getpid_noop.h:8]:
```suggestion
static OE_INLINE int oe_openssl_get_pid()
```
`static` since this function is not expected to be used elsewhere. `inline` opportunistically.

File: enclave/crypto/openssl/init.c [hunk 8]
```diff
@@ -15,6 +14,24 @@ int _is_symcrypt_engine_available = 0;
...
+static int _initialize_symcrypt_engine()
+{
+    int result;
+
+    result = SCOSSL_ENGINE_Initialize();
+    if (result != OE_SYMCRYPT_ENGINE_SUCCESS)
+        goto done;
+
+done:
+    if (result == OE_SYMCRYPT_ENGINE_FAIL)
```
Comment (anakrish) [inline enclave/crypto/openssl/init.c:26]: Can result be a value other than `OE_SYMCRYPT_ENGINE_FAIL` or `OE_SYMCRYPT_ENGINE_SUCCESS`? If so, error message won't always be printed.

**Review by anakrish — APPROVED** [review]
LGTM

**Comment by radhikaj** [comment]
@dcarpente , merging so I can continue building. Please address @anakrish's comments in the next PR. They are valuable and I am not ignoring them.

**Comment by radhikaj** [comment]
We have  tests failing - 994 - tests/openssl/3-evp_test-evppkey_kas (Child aborted)

	1137 - tests/crypto/enclave_openssl_3 (Failed)
